### PR TITLE
Feature - Pool advertisement backend mutations

### LINKED ISSUE #3168: Pool Advertisements - Backend Mutations
<sup>ℹ️ [Figma (root)][figma] | [Sitemap][sitemap]</sup>

## Pool Advertisements - Backend Mutations

Some changes to the backend model will need to be implemented to support the new pool advertisement pages.  Mutations will also need to be created.

- ARCHIVE is not included in this sprint

## Acceptance Criteria
- [x] Mutation for creating a pool advertisement [Wireframe for this bad boy](https://www.figma.com/proto/b6lDM86QCJQ6iaAh7qrTYH/Admin-v4-(Direct-Intake)?page-id=0%3A1&node-id=1656%3A5731&viewport=-3442%2C1747%2C0.59&scaling=min-zoom&starting-point-node-id=1618%3A6064)
- [x] Mutations for editing parts of a pool advertisement [Link to Edit Pool Advertisement wireframe](https://www.figma.com/proto/b6lDM86QCJQ6iaAh7qrTYH/Admin-v4-(Direct-Intake)?page-id=0%3A1&node-id=1618%3A6064&viewport=-3442%2C1747%2C0.59&scaling=min-zoom&starting-point-node-id=1618%3A6064)
- [x] Mutation for pool advertisement actions (publish, extend, etc)
- [x] Expiry date for "CREATE" has time of 23:59:59 Pacific Time
- [x] CREATE action sets pool owner to current user
- [x] "Close" action sets expiry date to "now"
- [x] Mutations are checking logic and returning user-viewable errors
  - [x] All mutations are ADMIN only
  - [x] Can only publish if in draft, all fields filled, expiry date in the future
  - [x] Can only delete if in draft

   [image]: https://user-images.githubusercontent.com/305339/166570935-6425a615-82c7-4b2f-9840-5c7d698fa2ab.png
   [image-link]: https://example.com
   [figma]: https://go.talent.c4nada.ca/figma
   [sitemap]: https://go.talent.c4nada.ca/sitemap

### PR DESCRIPTION
**Pull request by yonikid15**
Resolves #3168 

# Notes
- This PR creates new mutations for the Pool Advertisement pages. 
  - Create, Update, Publish, ExtendClosingDate, Close, and Delete.
- There is a `poolAdvertisementOperations.graphql` file in`talentsearch/src/js/api` with all the mutations which you use to test in the graphql-playground.

### CODE REVIEW

File: api/app/GraphQL/Mutations/ExtendPoolAdvertisementClosingDate.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+     */
+    public function __invoke($_, array $args)
+    {
+        $poolAdvertisement = Pool::find($args['id']);
+        $oldExpiryDate = $poolAdvertisement->expiry_date;
+        $newExpiryDate = $args['new_expiry_date'];
+
+        if ($newExpiryDate < Carbon::now()->endOfDay() ||
+            $newExpiryDate <= $oldExpiryDate) {
+            throw FormattedError::setInternalErrorMessage("New expiry date must be in the future and greater than the old expiry date.");
```
Comment (petertgiles) [inline api/app/GraphQL/Mutations/ExtendPoolAdvertisementClosingDate.php:24]: I wonder if this should be a more general "change expiry date" mutation that would allow the user to set any future date.  Requiring the new expiry date to be greater than the old expiry date seems not very useful for the API layer.

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -1184,6 +1184,48 @@ input GenericJobTitleBelongsToMany {
...
 
+input CreatePoolAdvertisementInput {
+    key: KeyString!
+    classifications: ClassificationBelongsToMany!
+}
+
+input UpdatePoolAdvertisementInput {
+    # Pool name and classification
+    name: LocalizedStringInput
+    classifications: ClassificationBelongsToMany!
```
Comment (petertgiles) [inline api/graphql/schema.graphql:1195]: Is there a reason why this is non-nullable?  It's a bit awkard to have to specify this in the query and I don't thing other update inputs have it non-nullable.

File: api/graphql/schema.graphql [hunk 3]
```diff
@@ -1184,6 +1184,48 @@ input GenericJobTitleBelongsToMany {
...
+    # Closing date
+    expiryDate: DateTimeTz @rename(attribute: "expiry_date")
```
Comment (petertgiles) [inline api/graphql/schema.graphql:1197]: This seems to be shifting the time to my local time zone.  For example, this:
```
{  
	"id": "389b2be0-0f2c-425d-8f99-eb2b4ff861fd",
  "poolAdvertisement": {
    "expiryDate": "2011-12-03T10:15:30+01:00",
    "classifications": { }
  }
}
```
Results in this:

image

I'd be OK if you wanted to kick this down the road to https://github.com/GCTC-NTGC/gc-digital-talent/issues/3315, too.

File: frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  $poolAdvertisement: UpdatePoolAdvertisementInput!
+) {
+  updatePoolAdvertisement(id: $id, poolAdvertisement: $poolAdvertisement) {
+    id
+  }
+}
+
+mutation publishPoolAdvertisement(
+  $id: ID!
+  $poolAdvertisement: PublishPoolAdvertisementInput!
```
Comment (petertgiles) [inline frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql:21]: What is the purpose of the $poolAdvertisement parameter?  Wouldn't you just need the ID for this action?

File: frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  extendPoolAdvertisementClosingDate(id: $id, newExpiryDate: $newExpiryDate) {
+    expiryDate
+  }
+}
+
+mutation closePoolAdvertisement($id: ID!) {
+  closePoolAdvertisement(id: $id) {
+    expiryDate
+  }
+}
```
Comment (petertgiles) [inline frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql:41]: I get this error when I try this mutation, and the extend mutation:
```
Add [expiry_date] to fillable property to allow mass assignment on [App\\Models\\Pool].
```

**Review by petertgiles — CHANGES_REQUESTED** [review]
Very cool Yoni!  This almost done, just a few things to check below.

File: api/graphql/schema.graphql [hunk 6]
```diff
@@ -1225,4 +1267,17 @@ type Mutation {
     deleteCommunityExperience(id: ID!): CommunityExperience @delete(globalId: false) @guard @can(ability: "delete", find: "id")
     deleteEducationExperience(id: ID!): EducationExperience @delete(globalId: false) @guard @can(ability: "delete", find: "id")
     deleteAwardExperience(id: ID!): AwardExperience @delete(globalId: false) @guard @can(ability: "delete", find: "id")
+
+    # PoolAdvertisement mutations
+    createPoolAdvertisement(userId: ID! @rename(attribute: "user_id"), poolAdvertisement: CreatePoolAdvertisementInput! @spread): PoolAdvertisement @create(model: "Pool") @guard @can(ability: "create", model: "Pool")
```
Comment (JamesHuf) [inline api/graphql/schema.graphql:1272]: I'm unsure of the use-cases for some of these mutations. For example:
When would we prefer this create mutation vs the createPool mutation when they do basically the same thing?
And why would we prefer each of these specific update mutations vs a more general one?

**Review by JamesHuf — CHANGES_REQUESTED** [review]
Things look good overall but I'm left with a couple questions.

**Comment by vd1992** [comment]
just checking things out, I see a bunch of good feedback above
I just wanted to mention this seems like something really important to test in some form, lots of tricky logic where some edge cases could creep into

File: api/app/GraphQL/Mutations/PublishPoolAdvertisement.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /**
+     * Publishes the pool advertisement.
+     * @param  null  $_
+     * @param  array{}  $args
+     */
+    public function __invoke($_, array $args)
+    {
+        $poolAdvertisement = Pool::find($args['id']);
+        $poolAdvertisement->update(['is_published' => true]);
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/GraphQL/Mutations/PublishPoolAdvertisement.php:20]: Yup! I created a [validation class](https://lighthouse-php.com/master/security/validation.html#validator-classes) which ties to the PublishPoolAdvertisementInput.

File: api/graphql/schema.graphql (hunk 2 above)
Comment (yonikid15) [reply api/graphql/schema.graphql:1195]: No this should be nullable!

File: api/app/GraphQL/Mutations/ExtendPoolAdvertisementClosingDate.php (hunk 1 above)
Comment (yonikid15) [reply api/app/GraphQL/Mutations/ExtendPoolAdvertisementClosingDate.php:24]: Ya I think you're right. I guess I was hooked on the word "extend", but having just a change date mutation makes more sense to me.

File: frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql (hunk 4 above)
Comment (yonikid15) [reply frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql:21]: Don't I want to validate the arguments that get passed in?

File: frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql (hunk 5 above)
Comment (yonikid15) [reply frontend/talentsearch/src/js/api/poolAdvertisementOperations.graphql:41]: Ah yes I need to add that to the fillable attribute in the Pool model.

File: api/app/GraphQL/Mutations/ClosePoolAdvertisement.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /**
+     * Closes the pool advertisements by setting the expiry_date to now().
+     * @param  null  $_
+     * @param  array{}  $args
+     */
+    public function __invoke($_, array $args)
+    {
+        $poolAdvertisement = Pool::find($args['id']);
+        $poolAdvertisement->update(['expiry_date' => Carbon::now()->endOfDay()]);
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/GraphQL/Mutations/ClosePoolAdvertisement.php:18]: Nope 😃. Thanks I'll fix that.

File: api/graphql/schema.graphql (hunk 6 above)
Comment (yonikid15) [reply api/graphql/schema.graphql:1283]: My understanding was to make explicit mutations for the new Pool Advertisement model and the events on the create and edit pages. However, you do have a point since a Pool Advertisement is just a Pool. @tristan-orourke Thoughts on this?

File: api/graphql/schema.graphql (hunk 3 above)
Comment (yonikid15) [reply api/graphql/schema.graphql]: Hmm, ya I'd like to get this merged in sooner then later since it's blocking other issues. Let's pass it onto #3315.

File: api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        $endOfDay = Carbon::now()->endOfDay();
+        return [
+            // Pool name and classification
+            'name.en' => [ 'required', 'string' ],
+            'name.fr' => [ 'required', 'string' ],
+            'classifications.sync' => [ 'required', 'array', 'min:1' ],
+            'classifications.sync.*' => [ 'required', 'uuid', 'exists:classifications,id' ],
+
+            // Closing date (Note: Form should return unzoned datetime.)
+            'expiryDate' => [ 'required', /*'date_format:Y-m-d H:i:s',*/ 'after:'.$endOfDay ], // TODO: Fix date_format validation
```
Comment (petertgiles) [inline api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php:28]: I see your notes here but I can't figure out a value that it will take.  I've tried a few different formats but no go.

image

**Review by petertgiles — CHANGES_REQUESTED** [review]
Nice, getting closer!  I've closed a few comments and opened a few more.

File: api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php (hunk 9 above)
Comment (yonikid15) [reply api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php:28]: Ya same here 😬 I was trying stuff from [PHP DateTime](https://www.php.net/manual/en/class.datetime.php)

File: api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            // Essential skills and Asset skills
+            'essentialSkills.sync' => [ 'required', 'array', 'min:1' ],
+            'essentialSkills.sync.*' => [ 'required', 'uuid', 'exists:skills,id' ],
+            'nonessentialSkills.sync' => [ 'required', 'array', 'min:1' ],
+            'nonessentialSkills.sync.*' => [ 'required', 'uuid', 'exists:skills,id' ],
+
+            // Other requirements
+            'advertisementLanguage' => [ 'required', Rule::in(ApiEnums::poolAdvertisementLanguages()) ],
+            'securityClearance' => [ 'required', Rule::in(ApiEnums::poolAdvertisementSecurity()) ],
+            'advertisementLocation.en' => [ 'required', 'string' ],
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php:45]: Ya should be possible with the [required_with](https://laravel.com/docs/9.x/validation#rule-required-with) rule

File: api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php (hunk 9 above)
Comment (vd1992) [reply api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php:28]: progress!

image

Comment (vd1992) [reply api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php:28]: it is broken though lol, the date check fails in PoolPolicy as a result

**Review by JamesHuf — APPROVED** [review]
I think this looks good beyond the points already mentioned.

**Comment by yonikid15** [comment]
Alright, I believe I got the rest of the notes Peter made. Unfortunately, I still haven't gotten the validation for the `date_format` to work properly. As well as showing the user friendly error messages 😠 which is proving to be more difficult then I thought.

File: api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php (hunk 9 above)
Comment (petertgiles) [reply api/app/GraphQL/Validators/PublishPoolAdvertisementInputValidator.php]: Got it!  I should have looked at the schema notes for the right format.  :roll_eyes: 
image

**Review by petertgiles — APPROVED** [review]
Nice!  I'm happy.  Yeah, the error validation messages suck but the frontend should be doing the heavy lifting on helping the user anyway.  I suggest spinning it off into a separate low-priority issue.

**Comment by vd1992** [comment]
for some reason, it only succeeds for me if I make this change

image

**Comment by vd1992** [comment]
> for some reason, it only succeeds for me if I make this change, I get "invalid data" otherwise
> 
> image

ah I see, that is intended for pools to be stuck to one Classification but the seeder currently attaches multiple Classifications to a pool instance

**Comment by vd1992** [comment]
I was playing around with validation, by making these changes (ignore the classifications one)

image

I got it to work I think

image

**Comment by vd1992** [comment]
From what I could tell, the date was being horribly mangled 

here

image

looking like this

image

**Comment by petertgiles** [comment]
> but the seeder currently attaches multiple Classifications to a pool instance

Good catch.  Might be worth updating that now.

**Comment by vd1992** [comment]
> > but the seeder currently attaches multiple Classifications to a pool instance
> 
> Good catch. Might be worth updating that now.

done here #3537
